graph/topo/clique_graph.go

```diff
@@ -0,0 +1,106 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package topo
+
+import (
+	"sort"
+
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/internal/ordered"
+	"gonum.org/v1/gonum/graph/internal/set"
+)
+
+// Builder is a pure topological graph construction type.
+type Builder interface {
+	AddNode(graph.Node)
+	SetEdge(graph.Edge)
+}
+
+// CliqueGraph builds the clique graph of g in dst using Clique and CliqueGraphEdge
+// nodes and edges. The nodes returned by calls to Nodes on the nodes and edges of
+// the constructed graph are the cliques and the common nodes between cliques
+// respectively. The dst graph is not cleared.
+func CliqueGraph(dst Builder, g graph.Undirected) {
+	cliques := BronKerbosch(g)
+
+	// Construct a consistent view of cliques in g. Sorting costs
+	// us a little, but not as much as the cliques themselves.
+	for _, c := range cliques {
+		sort.Sort(ordered.ByID(c))
+	}
+	sort.Sort(ordered.BySliceIDs(cliques))
+
+	cliqueNodes := make(cliqueNodeSets, len(cliques))
+	for id, c := range cliques {
+		s := make(set.Nodes, len(c))
+		for _, n := range c {
+			s.Add(n)
+		}
+		ns := &nodeSet{Clique: Clique{id: int64(id), nodes: c}, nodes: s}
+		dst.AddNode(ns.Clique)
+		for _, n := range c {
+			nid := n.ID()
+			cliqueNodes[nid] = append(cliqueNodes[nid], ns)
+		}
+	}
+
+	for _, cliques := range cliqueNodes {
+		for i, uc := range cliques {
+			for _, vc := range cliques[i+1:] {
+				// Retain the nodes that contribute to the
+				// edge between the cliques.
+				var edgeNodes []graph.Node
+				switch 1 {
```

Comment:
The first time I see this construct, wow.

Reply:
Yes, `switch {...` is just sugar for `switch true {...`. There are interesting an clever ways that you can abuse this.